### PR TITLE
Issue 990 slack today i learned

### LINKED ISSUE #990: Slack API: share a learning to CodeBuddies directly from Slack (@cb_jarvis)
Currently, users can schedule a hangout directly from Slack (see: https://github.com/codebuddies/codebuddies/pull/943). 

On the site, we have a learnings tracker that hasn't gotten a lot of love (only ~300 learnings reported):

screen shot 2018-10-02 at 2 55 54 am

As you might be able to tell from the screenshot, these learnings are typically shared on codebuddies.org/hangouts, underneath a hangout iframe on a hangout page, or in a study group tab. 

## TODO: 
1) Let users share a learning directly from Slack by typing in `@cb_jarvis #todayilearned [message]`

2) This message will appear as a learning under the `Learnings` collection, and be tied back to the user who submitted it.

screen shot 2018-10-02 at 2 59 20 am

3) Take a look at https://github.com/codebuddies/codebuddies/pull/943 for guidance on how to align a CodeBuddies Slack user email with the same user email on codebuddies.org.

### PR DESCRIPTION
**Pull request by railsstudent**
Fixes #990 

Create today I learned directly from slack by cb_jarvis bot.

@cb_jarvis til,  <Today I learned message>

screen shot 2018-12-09 at 9 52 01 pm
screen shot 2018-12-09 at 9 52 12 pm

### CODE REVIEW

**Comment by lpatmo** [comment]
Nice, you ended up writing the function that removed the extra <formatting>|https://formatting.com created by Slack!

**Comment by railsstudent** [comment]
@lpatmo Yes. However, I realized I have not replaced &lg; and &gt; with < and > respectively.  I will handle it if I can get off work on time today.

**Comment by railsstudent** [comment]
Added code to replace html encoded characters (&ampl;, &lt;, &gt;) to &, <, > respectively.

File: server/slack/webhooks.js [hunk 1]
```diff
@@ -141,6 +130,28 @@ const webhooks = {
...
+    }
+
+    LearningHelper.addLearning({
+      title: action.topic,
+      user_id: user._id,
+      username: user.username
+    });
+
+    const url = Meteor.absoluteUrl("/hangouts");
+    SlackAPI.postMessage(channel, `Today I learned is created for you at ${url}.`);
```
Comment (lpatmo) [inline server/slack/webhooks.js]: Could we change the message to `Your learning was archived on ${url}`?

File: server/slack/message-parser.js [hunk 2]
```diff
@@ -23,6 +48,20 @@ const Parser = {
       action.title = segments[2];
       action.date = Parser.getDate(segments[1]);
       if (!action.date) action.reply = "Date unrecognized. Please try Again.";
+    } else if (action.command === "til") {
+      const [_, ...rest] = segments;
+      const concatTopic = rest.join(", ").trim();
+      const topic = removeFormatting(concatTopic);
+      if (!topic) {
+        action.reply = "Please share something you learned. Try Again.";
```
[the remark this answers is not in the document]
Comment (stain88) [reply server/slack/message-parser.js]: should this hold for the "Again" in the date check above too?

File: server/slack/message-parser.js [hunk 3]
```diff
@@ -23,6 +48,20 @@ const Parser = {
       action.title = segments[2];
       action.date = Parser.getDate(segments[1]);
       if (!action.date) action.reply = "Date unrecognized. Please try Again.";
```
Comment (lpatmo) [inline server/slack/message-parser.js]: Ah yes, `Please try again` (no capitalization) here. Thanks for catching @stain88.

File: server/slack/message-parser.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (lpatmo) [reply server/slack/message-parser.js]: Yes, thanks for catching.

File: server/slack/webhooks.js (hunk 1 above)
Comment (railsstudent) [reply server/slack/webhooks.js]: sure

File: server/slack/message-parser.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (railsstudent) [reply server/slack/message-parser.js]: oops

File: server/slack/message-parser.js (hunk 3 above)
Comment (railsstudent) [reply server/slack/message-parser.js]: will fix

File: server/slack/webhooks.js (hunk 1 above)
Comment (railsstudent) [reply server/slack/webhooks.js]: @lpatmo  fixed

File: server/slack/message-parser.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (railsstudent) [reply server/slack/message-parser.js]: @lpatmo  fixed

File: server/slack/message-parser.js (hunk 3 above)
Comment (railsstudent) [reply server/slack/message-parser.js]: @lpatmo  fixed

**Comment by lpatmo** [comment]
@railsstudent Thank you! 🙏

**Pull request merged**